text/0745-ember-data-deprecate-methods-on-promise-many-array.md

```diff
@@ -0,0 +1,207 @@
+relationship results in changes to what records are available, those changes would not be
+reflected in the `commentIds`. More often, users will `await` the relationship to ensure the
+data is fully ready for use.
+
+```js
+const comments = await post.comments;
+const commentIds = comments.map(comment => comment.id);
+```
+
+As the language and framework have evolved, better features have become available that
+negate the value of many of the methods that exist on `PromiseManyArray`. This RFC seeks
+to better align the way folks interact with an async relationship with the `Octane` paradigm,
+`tracked` properties, and `async/await`. Doing so allows us to take a major first step towards
+simplifying the use of array-like proxies throughout EmberData.
+
+## Detailed design
+
+This RFC will reduce the API surface area of PromiseManyArray to just those methods needed to
+serve it's primary functions.
+
+- as a thenable/awaitable value (with promise state flags)
+- as a custom enumerable that for-convenience can be directly iterated by templates.
+
+All other methods and properties on PromiseManyArray would be deprecated. Additionally,
+the methods `addArrayObserver` and `removeArrayObserver` on `ManyArray` will be deprecated.
+
+Code examples, guides, and documentation should encourage users to resolve the relationship
+before operating on it, including avoiding the use of the remaining iterable APIs (which will
+be marked private).
+
+This deprecation would target `5.0` and would become `enabled` no-sooner than `4.1` (although
+it may be made `available` before then).
+
+We will keep the following properties and methods previously supplied by the `PromiseProxyMixin`
+
+- the public promise state flag `isSettled`
+- the public promise state flag `isPending`
+- the public promise state flag `isRejected`
+- the public promise state flag `isFulfilled`
+- the private property `promise`
+- the method `then`
+- the method `catch`
+- the method `finally`
+
+But we will eliminate the private property `reason`.
+
+We will keep the following property we had supplied on our own
+
+- the public property `links`
+- the method `reload`
+
+While deprecating the following methods which delegated to the ManyArray
+
+- `createRecord`
+
+As `PromiseManyArray` historically extended `ArrayProxy`, it inherited a large API surface from
+`MutableArray` `EmberArray` and `EmberObject` the totality of which will be deprecated *except* for:
+
+- the private property `content`
+- the property `length`
+- the public property `isDestroyed`
+- the iterator method `objectAt` (but which will now be marked private)
+
+
+### Finalized API of PromiseManyArray
+
+Remaining API surface
+
+- `content` (private)
+- `length`
+- `isDestroyed`
+- `objectAt` (private)
+- `links`
+- `reload`
+- `promise`  (private)
+- `isSettled`
+- `isPending`
+- `isFulfilled`
+- `isRejected`
+- `then`
+- `catch`
+- `finally`
+
+Deprecated API surface
+
```

Comment:
is [`toArray()`](https://api.emberjs.com/ember/4.1/classes/ArrayProxy/methods/toArray?anchor=toArray) missing here?

Reply:
yes, thanks for the catch!